syntax = "proto3";

package duckdb.distributed;

// Request types for distributed operations
enum RequestType {
  REQUEST_TYPE_UNSPECIFIED = 0;
  EXECUTE_SQL = 1;
  CREATE_TABLE = 2;
  DROP_TABLE = 3;
  INSERT_DATA = 4;
  SCAN_TABLE = 5;
  DELETE_DATA = 6;
  TABLE_EXISTS = 7;
}

// Execute SQL request
message ExecuteSQLRequest {
  string sql = 1;
}

// Create table request
message CreateTableRequest {
  string sql = 1;
}

// Drop table request
message DropTableRequest {
  string table_name = 1;
}

// Scan table request
message ScanTableRequest {
  string table_name = 1;
  uint64 limit = 2;
  uint64 offset = 3;
}

// Delete data request
message DeleteDataRequest {
  string table_name = 1;
  string where_clause = 2;
}

// Table exists request
message TableExistsRequest {
  string table_name = 1;
}

// Request message for distributed operations
message DistributedRequest {
  // Request-specific parameters
  oneof request {
    ExecuteSQLRequest execute_sql = 1;
    CreateTableRequest create_table = 2;
    DropTableRequest drop_table = 3;
    ScanTableRequest scan_table = 4;
    DeleteDataRequest delete_data = 5;
    TableExistsRequest table_exists = 6;
  }

  // Common metadata for all requests
  map<string, string> metadata = 10;
}

// Execute SQL response
message ExecuteSQLResponse {
  uint64 rows_affected = 1;
}

// Create table response
message CreateTableResponse {
  // Empty - success indicated by parent
}

// Drop table response
message DropTableResponse {
  // Empty - success indicated by parent
}

// Scan table response (data returned via Arrow Flight)
message ScanTableResponse {
  uint64 row_count = 1;
}

// Delete data response
message DeleteDataResponse {
  uint64 rows_deleted = 1;
}

// Table exists response
message TableExistsResponse {
  bool exists = 1;
}

// Response message for distributed operations
message DistributedResponse {
  // Success or error
  bool success = 1;
  string error_message = 2;

  // Response-specific data
  oneof response {
    ExecuteSQLResponse execute_sql = 3;
    CreateTableResponse create_table = 4;
    DropTableResponse drop_table = 5;
    ScanTableResponse scan_table = 6;
    DeleteDataResponse delete_data = 7;
    TableExistsResponse table_exists = 8;
  }

  // Common metadata for all responses
  map<string, string> metadata = 10;
}

// Service definition for distributed operations
// This integrates with Arrow Flight for data transfer
service DistributedExecutor {
  // Execute a control operation (uses protobuf)
  rpc ExecuteControl(DistributedRequest) returns (DistributedResponse);

  // Note: Data transfer uses Arrow Flight's DoGet/DoPut
  // instead of protobuf for efficiency
}
